[feat]: add transcription API endpoint using OpenAI Whisper-small

davidgao7 opened the pull request:
**\[Feat] Add transcription API endpoint using OpenAI Whisper-small**

This PR implements a new `/v1/audio/transcriptions` route in the vllm-router so you can upload audio files directly through the router and have them forwarded to a Whisper transcription backend. It includes:

- **Endpoint**: `/v1/audio/transcriptions`
- **Parameters**:  
  - `file` (UploadFile)  
  - `model` (e.g. `openai/whisper-small`)  
  - `prompt` (optional)  
  - `response_format` (`json`, `text`, `srt`, `verbose_json`, or `vtt`)  
  - `temperature` (optional)  
  - `language` (default `"en"`)
- **Behavior**  
  - Reads incoming bytes, rebuilds a `multipart/form-data` payload
  - Discovers transcription backends via service discovery 
  - Proxies requests (with extended HTTPX timeouts) to the Whisper node  
  - Returns the backend’s raw response unmodified  
- **Script**: `run-router.sh` to spin up a static‐mode router pointing at a local Whisper server   

**Fixes** [#410](https://github.com/vllm-project/production-stack/issues/410)

**Testing**

- **Local / RunPod (static mode)**  
  1. Start Whisper backend (RunPod):  
     ```bash
     vllm serve \
       --task transcription openai/whisper-small \
       --host 0.0.0.0 --port 8002
     ```
  2. Start router against it on RunPod:  
     ```bash
     ./run-router.sh 8000 http://localhost:8002
     ```
  3. Curl a test file:  
     ```bash
     curl -v http://localhost:8000/v1/audio/transcriptions \
       -F 'file=@/path/to/test.wav;type=audio/wav' \
       -F 'model=openai/whisper-small' \
       -F 'response_format=json' \
       -F 'language=en'
     ```
     → 200 OK with JSON transcript:
    ```json
    {
        "text": "This is a test transcription."
    }
    ```

- **Kubernetes (production mode)**  
  Switch to `--service-discovery k8s` (or dynamic JSON config) so real vLLM pods are picked up automatically instead of the static backend.

---

**Checklist**

* [x] Code changes pass `pre-commit run --all-files`.
* [x] Signed-off my commits with `-s`.
* [x] Classified this PR as `[Feat]`.
*  [ ] Documentation updated (if needed)  
*  [ ] Integration or unit tests added (if applicable)  
<details>  
<summary><b> Detailed Checklist (Click to Expand) </b></summary>

<p>Before submitting, please ensure the PR meets the following:</p>

1. **PR Title and Classification**

   * Prefixed with `[Feat]` since this introduces a new feature.
2. **Code Quality**

   * All linter/formatter checks pass (run `pre-commit run --all-files`).
   * Code is well-documented and clear.
   * Added or updated tests to cover the new transcription endpoint.
3. **DCO & Signed-off**

   * Commits include a `Signed-off-by:` trailer agreeing to the DCO.
   * Used `git commit -s` for automatic sign-off.

</details>

nmiguel commented:
When I run check out this branch to test the transcriptions API, I get the following error:

```
File "/usr/local/lib/python3.12/site-packages/vllm_router/routers/main_router.py", line 248, in audio_transcriptions
    if model in ep.model_names  # that actually serve your model
                ^^^^^^^^^^^^^^
AttributeError: 'EndpointInfo' object has no attribute 'model_names'. Did you mean: 'model_name'?
```
It seems that EndpointInfo only has a `model_name: str` parameter, and not the `model_names:list[str]` that this PR expects.

davidgao7 commented:
I’ve put together the steps I followed [here](https://davidgao7.github.io/posts/vllm-v1-whisper-transcription/)](https://davidgao7.github.io/posts/vllm-v1-whisper-transcription/)  -- would love any feedback if I missed something or got anything wrong! 😊

YuhanLiu11 commented:
@davidgao7 This is super awesome! I'll take a look

YuhanLiu11 commented:
Hey @davidgao7 this is an awesome feature to have! A general comment: can you remove all the debugging messages in the code?

max-wittig commented on src/vllm_router/utils.py:
```diff
@@ -75,6 +76,10 @@ def get_test_payload(model_type: str):
                 return {"encoding_format": "float", "text_1": "Test", "test_2": "Test2"}
+            case ModelType.transcription:
+                return {
+                    "file": "",
```
This is used to test the payload so ideally we would have an example audio file here. Maybe we can generate a very short one on application startup?

davidgao7 replied:
> This is used to test the payload so ideally we would have an example audio file here. Maybe we can generate a very short one on application startup?
> or does it also work with an empty file?

Hi Max,

Thanks for the suggestion! I’m planning to go with the empty audio file approach first — the idea is to pass an empty in-memory file (BytesIO()) as the payload to test compatibility with the transcription endpoint. If that works reliably, it’ll keep things clean without needing to generate or manage any example files.

I’ll test it out today and report back on whether it behaves as expected. If it doesn’t work well, we can revisit the idea of generating a short clip at startup.

Let me know if you have any thoughts before I proceed!

max-wittig replied:
If you want to test, if the test payload works by using `--static-backend-health-checks`.

For this to work, you have to go via the normal endpoint health API though: https://github.com/vllm-project/production-stack/blob/main/src/vllm_router/service_discovery.py#L282

If you have any questions, feel free to just ask and I can try to help as well!

davidgao7 commented:
> @davidgao7 Hi David, would you mind fixing the pre-commit check issue and the review comment mentioned above?

Hi @zerofishnoodles — thanks for the reminder! Sorry for the delay, I’ve been a bit tied up but finally have some solid time blocks again and can jump back in. I’ll fix the pre-commit issue and address Max’s review comment shortly!

max-wittig commented on src/vllm_router/utils.py:
```diff
@@ -75,6 +78,22 @@ def get_test_payload(model_type: str):
                 return {"query": "Hello", "documents": ["Test"]}
             case ModelType.score:
                 return {"encoding_format": "float", "text_1": "Test", "test_2": "Test2"}
+            case ModelType.transcription:
+                # Generate a 0.1 second silent audio file
+                with io.BytesIO() as wav_buffer:
+                    with wave.open(wav_buffer, "wb") as wf:
+                        wf.setnchannels(1)  # mono audio channel, standard configuration
+                        wf.setsampwidth(2)  # 16 bit audio, common bit depth for wav file
+                        wf.setframerate(16000)  # 16 kHz sample rate
+                        wf.writeframes(b"\x00\x00" * 1600)  # 0.1 second of silence
+
+                    # retrieves the generated wav bytes, return
+                    wav_bytes = wav_buffer.getvalue()
+
+                return {
+                    "file": ("empty.wav", wav_bytes, "audio/wav"),
+                }
+
```
this is very cool! 

Minor comment, feel free to ignore: Could we add this as a constant somewhere to avoid the bytebuffer to be recreated every minute?

davidgao7 replied:
@max-wittig Hi Max,

Quick update on the silent WAV bytes. I got it implemented as a module-level constant (`_SILENT_WAV_BYTES`) in [src/vllm_router/utils.py](https://github.com/davidgao7/production-stack/blob/4cf9182062d3be848ee8ca06ddb1e88d01400e14/src/vllm_router/utils.py#L17C1-L30C1), so that's all squared away on preventing repeated creation, which is great!

While I was digging into the health checks for the `transcription` model, I noticed an interesting log pop up, like this one:

<img width="609" alt="Screenshot 2025-06-19 at 4 49 56 PM" src="https://github.com/user-attachments/assets/e151e5c4-78ee-4df5-be50-5e8555612831" />

It looks like the `is_model_healthy` function ([line 185 in utils.py](https://github.com/davidgao7/production-stack/blob/4cf9182062d3be848ee8ca06ddb1e88d01400e14/src/vllm_router/utils.py#L185)) is designed to send a JSON payload ([around line 190](https://github.com/davidgao7/production-stack/blob/4cf9182062d3be848ee8ca06ddb1e88d01400e14/src/vllm_router/utils.py#L190)) for health checks, which works perfectly for most models!

**The core issue is this:**
The `transcription` endpoint (`/v1/audio/transcriptions`) on the Whisper backend specifically expects `multipart/form-data` with an audio file – just like how the main transcription endpoint in `main_router.py` is designed to handle user uploads (it uses `UploadFile` and builds `files=files`). So, when the health check sends JSON, the backend doesn't quite know what to do with it and rejects the request, making the health check show up as `unhealthy`.

I was thinking a good way to get this working smoothly for `transcription` models would be to adjust the `is_model_healthy` function to specifically send a minimal `multipart/form-data` request, using our `_SILENT_WAV_BYTES`. This would let us correctly mimic a real audio upload for the health check.

Here’s roughly what that change in `src/vllm_router/utils.py` might look like:

```python
# In is_model_healthy function in utils.py
def is_model_healthy(url: str, model: str, model_type: str) -> bool:
    model_details = ModelType[model_type]
    try:
        if model_type == "transcription":
            # For transcription, the backend expects multipart/form-data with a file.
            # We'll use our pre-generated silent WAV bytes.
            files = {
                'file': ('silent.wav', _SILENT_WAV_BYTES, 'audio/wav')
            }
            # The model name also needs to be sent as part of the form data
            data = {'model': model}
            response = requests.post(
                f"{url}{model_details.value}",
                files=files,  # This sends multipart/form-data
                data=data     # And this includes other form fields
            )
        else:
            # Existing logic for other model types (chat, completion, etc.)
            response = requests.post(
                f"{url}{model_details.value}",
                headers={"Content-Type": "application/json"},
                json={"model": model} | model_details.get_test_payload()
            )

        response.raise_for_status()  # Throws an error for 4xx/5xx responses
        # For transcription, we just need to confirm a 200 OK.
        # Other models might need to parse JSON.
        if model_type == "transcription":
            return True
        else:
            response.json() # Verify it's valid JSON for other model types
        return True
    except requests.exceptions.RequestException as e:
        logger.warning(f"{model_type} model {model} at {url} not healthy: {e}")
        return False
    except json.JSONDecodeError as e:
        logger.error(f"Failed to decode JSON from {model_type} model {model} at {url}: {e}")
        return False

# And then ModelType.transcription.get_test_payload() could just return {}
# since it wouldn't be used by this new multipart path:
# In ModelType class in utils.py
class ModelType(enum.Enum):
    chat = "/v1/chat/completions"
    completion = "/v1/completions"
    embeddings = "/v1/embeddings"
    rerank = "/v1/rerank"
    score = "/v1/score"
    transcription = "/v1/audio/transcriptions"

    @staticmethod
    def get_test_payload(model_type: str):
        match ModelType[model_type]:
            case ModelType.chat:
                return {
                    "messages": [
                        {
                            "role": "user",
                            "content": "Hello",
                        }
                    ],
                    "temperature": 0.0,
                    "max_tokens": 3,
                    "max_completion_tokens": 3,
                }
            case ModelType.completion:
                return {"prompt": "Hello"}
            case ModelType.embeddings:
                return {"input": "Hello"}
            case ModelType.rerank:
                return {"query": "Hello", "documents": ["Test"]}
            case ModelType.score:
                return {"encoding_format": "float", "text_1": "Test", "test_2": "Test2"}
            case ModelType.transcription:
                # This payload is for JSON part of the request.
                # The file bytes will be handled separately via 'files=' parameter
                # in the is_model_healthy function for transcription health checks.
                return {} # Changed back to an empty dictionary


    @staticmethod
    def get_all_fields():
        return [model_type.name for model_type in ModelType]
```

What do you think about this approach? I believe it should get the health checks reliably working for transcription models!

zerofishnoodles commented on src/vllm_router/routers/main_router.py:
```diff
@@ -232,3 +244,123 @@ async def health() -> Response:
+    router = get_routing_logic()
+
+    # pick one using the router's configured logic (roundrobin, least-loaded, etc.)
+    chosen_url = router.route_request(
```
Would it be better to put the routing logic like getting the endpoint, routing the endpoint and proxy to the request service to keep the main_router concise?

zerofishnoodles commented on src/vllm_router/routers/main_router.py:
```diff
@@ -232,3 +244,123 @@ async def health() -> Response:
+    response_format: str | None = Form("json"),
+    temperature: float | None = Form(None),
+    language: str = Form("en"),
+):
```
Would it be better to align the parameter with the other endpoints? And also perhaps adding request-id would be a good one.

zerofishnoodles reviewed (commented):
Could you please take a little time to fix the GitHub check issue?

huiwq1990 commented:
I think it's better reuse `vllm_router.services.request_service.request.route_general_request` to process the request.

davidgao7 commented:
> I think it's better reuse `vllm_router.services.request_service.request.route_general_request` to process the request.

This is helpful, @huiwq1990! I'll read through your implementation carefully and plan to combine the insights into my existing code.

max-wittig reviewed (changes requested):
Some comments

davidgao7 commented:
Hi @max-wittig,

Hope you're well!

* I've cleaned up and updated the comments for the `_SILENT_WAV_BYTES` generation in `utils.py` for clarity.
* All relevant files have now passed auto-formatting and linting checks via `pre-commit`, ensuring consistent project style. (This also addresses the import formatting question from before.)

Please let me know your thoughts when you get a chance! Thanks again for your review!

max-wittig reviewed (approved):
Looks good, so over to the actual maintainers

Maybe you could take another look @YuhanLiu11 ?

zerofishnoodles commented:
Hi @davidgao7 Can you fix the CI check issues and merge base branch? And then I think we are good to go.

max-wittig commented:
Hi @davidgao7. Please run isort for the pre-commit checks to succeed

davidgao7 commented:
> Hi @davidgao7. Please run isort for the pre-commit checks to succeed

I have committed the change applied on pre-commit.

Should I commit `uv.lock`? 

ref: https://github.com/astral-sh/uv/issues/9797

max-wittig commented:
@davidgao7 I guess this is out of scope for this MR.

FuxMak commented:
Just stumbled about this during the implementation of a STT model. There has been a lot of work put into from both sides, thanks a lot!

davidgao7 commented:
Hi @max-wittig,

* I've added `httpx` as an explicit dependency to `pyproject.toml` and `requirements.txt`, which should fix the `ModuleNotFoundError` the CI was hitting.
* I also ran `pre-commit run --all-files` to fix the `isort` and other formatting issues, so those checks should now pass.

The branch should be ready for another workflow run when you have a moment. I'll keep an eye on the results and am ready for any further feedback.

Thanks!

jonoillar commented:
Hi @davidgao7 , starting from this MR: https://github.com/vllm-project/production-stack/pull/589 `httpx` has been replaced by `aiohttp`. The reason is a huge performance boost

So I'd advise not to re introduce `httpx`, and rather use `aiohttp`, which is already part of the package requirements

zerofishnoodles commented:
Hi, can you fix the pre-commit issue? Also a suggestion, it would be better to install the git pre-commit and also run pre-commit manually before committing.

max-wittig commented:
@YuhanLiu11 Could you take a look at this? I think this has been open so long by now and it would provide a big benefit for the people.

YuhanLiu11 commented:
> @YuhanLiu11 Could you take a look at this? I think this has been open so long by now and it would provide a big benefit for the people.

Yeah I just started running the CI test and let's see if it passes the tests.

max-wittig commented:
Thank you all for this journey! We will test it!
